### PR TITLE
Add Category management

### PR DESCRIPTION
**Pull request by ShastkivRuslan**
This pull request introduces new functionality for managing categories within the project, including role-based access control for users (USER) and administrators (ADMIN), full CRUD operations for categories, and integration with existing book data. Below is an overview of the implemented changes:
Key Changes:

1. Added Category entity with fields: id (PK), name (not null), and description.
2. Updated Book entity to include a Set<Category> categories field to support categorization.
3. Created CategoryRepository extending JpaRepository for database access.
4. Implemented DTOs for Category entity, enabling clean and organized data transfer.
5. Enhanced BookMapper with methods:

    - BookDto toDto(Book book);
    - Book toEntity(CreateBookRequestDto bookDto);
    - BookDtoWithoutCategoryIds toDtoWithoutCategories(Book book);

6. Created CategoryMapper with methods:

      - CategoryDto toDto(Category category);
      - Category toEntity(CategoryDto categoryDTO);
      - void updateCategoryFromDto(CategoryRequestDto requestDto, @MappingTarget Category category);

7. Developed CategoryService interface and CategoryServiceImpl with:

    - Page<CategoryDto> findAll()
    - CategoryDto getById(Long id)
    - CategoryDto save(CategoryDto categoryDto)
    - CategoryDto update(Long id, CategoryDto categoryDto)
    - void deleteById(Long id)

8. Created CategoryController with endpoints:

      - POST /api/categories - createCategory(CategoryDto categoryDto)
      - GET /api/categories - getAll()
      - GET /api/categories/{id} - getCategoryById(Long id)
      - PUT /api/categories/{id} - updateCategory(Long id, CategoryDto categoryDto)
      - DELETE /api/categories/{id} - deleteCategory(Long id)
      - GET /api/categories/{id}/books - getBooksByCategoryId(Long id)
      
9. Implemented search functionality by category IDs, allowing users to filter books based on selected categories
10. Added liquibase scripts to create 'category', 'book_category' tables, and insert test data into these tables.

This PR also includes updates to mappers, DTOs, and the Book entity to support the category relationship, allowing books to be categorized effectively in the application.

### CODE REVIEW

File: src/main/java/ruslan/shastkiv/bookstore/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+    @GetMapping("/{id}")
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "Get category by ID",
+            description = "This endpoint allows users to fetch a category by its ID.")
+    public CategoryDto getCategoryById(@PathVariable Long id) {
+        return categoryService.getCategoryById(id);
+    }
+
+    @GetMapping("/{id}/books")
```
Comment (ahoienko) [inline src/main/java/ruslan/shastkiv/bookstore/controller/CategoryController.java:61]: PreAuthorize is missing

File: src/main/java/ruslan/shastkiv/bookstore/service/book/BookServiceImpl.java [hunk 2]
```diff
@@ -56,9 +64,22 @@ public Page<BookDto> search(BookSearchParametersDto searchParametersDto, Pageabl
...
     private Book findBookById(Long id) {
         return bookRepository.findById(id).orElseThrow(()
                 -> new EntityNotFoundException("Can t find book by id: " + id));
     }
 
+    private Set<Category> initCategories(List<Long> categoryIds) {
+        return categoryIds.stream()
+                .map(id -> categoryRepository.findById(id).orElseThrow(()
+                        -> new EntityNotFoundException("Cant find category by id: " + id)))
+                .collect(Collectors.toSet());
```
Comment (ahoienko) [inline src/main/java/ruslan/shastkiv/bookstore/service/book/BookServiceImpl.java]: What if you want to set 1000 categories to book? You will make a DB request to get each separately this is not efficient. so please think how to such operation using 1 request only

File: src/main/java/ruslan/shastkiv/bookstore/service/book/BookService.java [hunk 3]
```diff
@@ -19,4 +20,6 @@ public interface BookService {
 
     Page<BookDto> search(BookSearchParametersDto searchParametersDto, Pageable pageable);
 
+    Page<BookDtoWithoutCategoryIds> getAllBooksByCategoryId(Long id, Pageable pageable);
 }
+
```
Comment (ahoienko) [inline src/main/java/ruslan/shastkiv/bookstore/service/book/BookService.java]:
```suggestion
```

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 4]
```diff
@@ -13,3 +13,14 @@ databaseChangeLog:
...
+      file: db/changelog/changes/08-create-categories-table.yaml
+  - include:
+      file: db/changelog/changes/09-create-books_categories-table.yaml
+  - include:
+      file: db/changelog/changes/10-insert-categories.yaml
+  - include:
+      file: db/changelog/changes/11-insert-books.yaml
+  - include:
+      file: db/changelog/changes/12-insert-books_categories.yaml
+
```
Comment (ahoienko) [inline src/main/resources/db/changelog/db.changelog-master.yaml]:
```suggestion
```

File: src/main/resources/db/changelog/changes/12-insert-books_categories.yaml [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+      changes:
+        - insert:
+            tableName: books_categories
+            columns:
+              - column:
+                  name: book_id
+                  value: "1"
+              - column:
+                  name: category_id
+                  value: "1"
```
Comment (ahoienko) [inline src/main/resources/db/changelog/changes/12-insert-books_categories.yaml]: Better to use valueComputed to make sure that you made correct relation

File: src/main/resources/db/changelog/changes/09-create-books_categories-table.yaml [hunk 6]
```diff
@@ -0,0 +1,26 @@
+databaseChangeLog:
+  - changeSet:
+      id: create_books_categories_table
+      author: shastkiv
+      changes:
+        - createTable:
+            tableName: books_categories
+            columns:
+              - column:
+                  name: book_id
```
Comment (ahoienko) [inline src/main/resources/db/changelog/changes/09-create-books_categories-table.yaml:10]: Id is also unique, I assume that there is some PK constraints that includes all necessary constraints please check it

**Review by ahoienko — CHANGES_REQUESTED** [review]
Nice, left some comments)

File: src/main/java/ruslan/shastkiv/bookstore/service/book/BookServiceImpl.java (hunk 2 above)
Comment (ShastkivRuslan) [reply src/main/java/ruslan/shastkiv/bookstore/service/book/BookServiceImpl.java]: I agree with you, but I was also thinking about how to notify the admin which category is missing in the database.

**Comment by ShastkivRuslan** [comment]
Changes after mentor review
- added @JsonIgnore annotation to the 'isDeleted' field in Category entity to exclude it from being serialized when loading BookDto with categories;
- refactored initCategories method to load all categories at once for better performance;
- added logic to check and throw an exception with the specific category id if a non-existent category id is provided during book creation:
- updated 12-insert-books_categories.yaml to use valueComputed for the book-category relation to ensure correctness of the relation.

File: src/main/java/ruslan/shastkiv/bookstore/controller/CategoryController.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+@Tag(name = "Category", description = "Operations related to categories")
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @PostMapping
+    @PreAuthorize("hasRole('ADMIN')")
+    @Operation(summary = "Add a new category",
+            description = "This endpoint allows an admin to add a new category.")
+    public CategoryDto addCategory(@RequestBody @Valid CategoryRequestDto requestDto) {
```
Comment (JJJazl) [inline src/main/java/ruslan/shastkiv/bookstore/controller/CategoryController.java:41]: Please, add an appropriate status code for POST method

File: src/main/java/ruslan/shastkiv/bookstore/model/Category.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+public class Category {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false)
+    private String name;
+    private String description;
+    @Column(nullable = false)
+    @JsonIgnore
+    private boolean isDeleted = false;
```
Comment (JJJazl) [inline src/main/java/ruslan/shastkiv/bookstore/model/Category.java:30]: Why do we need @JsonIgnore for this field? We don't want to return a `Category ` model from our controllers as a response or part of it

File: src/main/java/ruslan/shastkiv/bookstore/dto/book/CreateBookRequestDto.java [hunk 9]
```diff
@@ -28,4 +31,7 @@ public class CreateBookRequestDto {
 
     @CoverImage
     private String coverImage;
+
+    @NotEmpty(message = "The book must belong to at least one category.")
+    private List<Long> categoryIds = new ArrayList<>();
```
Comment (JJJazl) [inline src/main/java/ruslan/shastkiv/bookstore/dto/book/CreateBookRequestDto.java]:
```suggestion
    private Set<Long> categoryIds = new HashSet<>();
```

File: src/main/java/ruslan/shastkiv/bookstore/service/book/BookServiceImpl.java [hunk 10]
```diff
@@ -56,9 +65,30 @@ public Page<BookDto> search(BookSearchParametersDto searchParametersDto, Pageabl
...
+        return bookRepository.findBooksByCategoryId(id, pageable)
+                .map(bookMapper::toDtoWithoutCategoryIds);
+    }
+
     private Book findBookById(Long id) {
         return bookRepository.findById(id).orElseThrow(()
                 -> new EntityNotFoundException("Can t find book by id: " + id));
     }
 
+    private Set<Category> initCategories(List<Long> categoryIds) {
```
Comment (JJJazl) [inline src/main/java/ruslan/shastkiv/bookstore/service/book/BookServiceImpl.java]:
```suggestion
    private Set<Category> initCategories(Set<Long> categoryIds) {
```

File: src/main/java/ruslan/shastkiv/bookstore/service/book/BookServiceImpl.java [hunk 11]
```diff
@@ -56,9 +65,30 @@ public Page<BookDto> search(BookSearchParametersDto searchParametersDto, Pageabl
...
+        List<Category> categoriesByIds = categoryRepository.findAllById(categoryIds);
+
+        Set<Long> checkList = categoriesByIds.stream()
+                .map(Category::getId)
+                .collect(Collectors.toSet());
```
Comment (JJJazl) [inline src/main/java/ruslan/shastkiv/bookstore/service/book/BookServiceImpl.java]:
```suggestion
        Set<Long> categoryIdsInDB = categoriesByIds.stream()
                .map(Category::getId)
                .collect(Collectors.toSet());
```

**Review by JJJazl — CHANGES_REQUESTED** [review]

File: src/main/java/ruslan/shastkiv/bookstore/service/book/BookServiceImpl.java [hunk 12]
```diff
@@ -56,9 +65,30 @@ public Page<BookDto> search(BookSearchParametersDto searchParametersDto, Pageabl
...
+
+        for (Long id : categoryIds) {
+            if (!categoryIdsInDB.contains(id)) {
+                throw new EntityNotFoundException("Can`t find category by id: " + id);
+            }
+        }
+        return new HashSet<>(categoriesByIds);
+    }
+
```
Comment (okuzan) [inline src/main/java/ruslan/shastkiv/bookstore/service/book/BookServiceImpl.java]:
```suggestion
```

File: src/main/java/ruslan/shastkiv/bookstore/model/Category.java [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Entity
+@Table(name = "categories")
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE categories SET is_deleted = TRUE WHERE id =?")
+@SQLRestriction("is_deleted=false")
```
Comment (okuzan) [inline src/main/java/ruslan/shastkiv/bookstore/model/Category.java]:
```suggestion
@SQLDelete(sql = "UPDATE categories SET is_deleted = TRUE WHERE id = ?")
@SQLRestriction("is_deleted = FALSE")
```

**Review by okuzan — APPROVED** [review]

**Pull request merged**